changes to episode 4, tasks 17-27 (#143)
carpentries-incubator/high-dimensional-stats-r: _episodes_rmd/04-principal-component-analysis.Rmd

more useful at explaining your data than others. By definition, the first
principal component explains more variation than other principal components.

> ```{r fig1, echo=FALSE, fig.cap="Scatter plots of the farm data with the first principal component in red and second principal component in green.", fig.alt="Side-by-side scatter plots of the farm data. The left figure displays a scatter plot of biodiversity versus percentage area fallow. The first principal component is overlain and shown by a red line and the second principal component is shown by a green line. The right figure displays the same scatter plot rotated so that the first principal component is horizontal and the second principal component is shown perpendicular to this."}

Reviewer: ```suggestion
```{r fig1, echo=FALSE, fig.cap="Scatter plots of the farm data with the first principal component in red and second principal component in green.", fig.alt="Side-by-side scatter plots of the farm data. The left figure displays a scatter plot of biodiversity versus percentage area fallow. The first principal component is shown by a red line and the second principal component is shown by a green line. The right figure displays the same scatter plot rotated so that the first principal component is horizontal and the second principal component is shown perpendicular to this."}
```

principal component.
This is explained in more detail on [this Q&A website](https://stats.stackexchange.com/questions/2691/making-sense-of-principal-component-analysis-eigenvectors-eigenvalues).

> ```{r pendulum, echo=FALSE, fig.cap="Animation showing the iterative process by which principal components are found. Data points are shown in blue, distances from the line in red dashed lines and the principal component is shown by a black line.", fig.alt="Animated scatter plot displaying the iterative process by which principal components are found. The data points are shown in blue and the principal component is shown by a solid black line. The distances of the points from the line are also shown by red dashed lines. The animation initially starts with the principal component far away from the direction of variability in the points and, as time goes on, eventually finds the direction of variability exhibiting a springing motion."}

Reviewer: ```suggestion
```{r pendulum, echo=FALSE, fig.cap="Animation showing the iterative process by which principal components are found. Data points are shown in blue, distances from the line as red dashed lines and the principal component is shown by a black line.", fig.alt="Animated scatter plot displaying the iterative process by which principal components are found. The data points are shown in blue and the principal component is shown by a solid black line. The distances of the points from the line are shown by red dashed lines. The animation initially starts with the principal component far away from the direction of variability in the points and, as time goes on, eventually finds the direction of variability exhibiting a springing motion."}
```

```

```{r, echo = FALSE}

> # Get proportions of variance explained by each principal component (rounded to 2 DP)

Reviewer: ```suggestion
# Get proportions of variance explained by each principal component (rounded to 2 decimal places)
```

```

```{r vardf-plot, fig.cap="Scree plot showing the percentage of the variance explained by the principal components calculated from the prostate data.", fig.alt="A scree plot showing the percentage of variance explained by each principal component versus the principal component number. The points are joined by lines to indicate where the elbow of the scree plot occurs."}
plot(varDF, type="b")
```

The scree plot shows that the first principal component explains most of the
variance in the data (>50%) and each subsequent principal component explains
less and less of the total variance. The first two principal components
explain >70% of variance in the data. 

Selecting a subset of principal components results in loss of 
information from the original dataset but selecting principal components to 
summarise the _majority_ of the information in the original dataset is useful
from a practical dimension reduction perspective. There are no clear guidelines 
on how many principal components should be included in PCA. 
We often look at the 'elbow’ on the scree plot as an
indicator that the addition of principal components does not drastically

> contribute to explain the remaining variance. In this case, the 'elbow' of the scree plot appears

Reviewer: ```suggestion
contribute to explaining the remaining variance. In this case, the 'elbow' of the scree plot appears
```

to be at the third principal component. We may therefore conclude that three principal
components are sufficient to explain the majority of the variability in the original 
dataset. We may also choose another criterion, such as an arbitrary cut-off for 
the proportion of the variance explained compared to other principal components. 
Essentially, the criterion used to select principal components should be determined 
based on what is deemed a sufficient level of information retention for a specific 

> dataset.

Reviewer: ```suggestion
dataset and question.
```

> ## Computing a PCA "by hand"
> The rotation matrix obtained in a PCA is identical to the eigenvectors
> of the covariance matrix of the data. Multiplying these with the (centered and scaled)
> data yields the principal component scores:
> ```{r pca-by-hand, fig.cap="Scatter plots of the second versus first principal components calculated by prcomp() (left) and by hand (right).", fig.alt="Side-by-side scatter plots of the second versus first principal components calculated by prcomp() (left) and by hand (right). The left scatter plot is the same as the right scatter plot but with swapped axes."}
> pros2.scaled <- scale(pros2) # centre and scale the prostate data

> > pros2.cov <- cov(pros2.scaled)   #generate covariance matrix

Reviewer: ```suggestion
> pros2.cov <- cov(pros2.scaled)   # generate covariance matrix
```

the desired package when calling `biplot()`. A biplot of the first two principal
components can be generated as follows:

> ```{r stats-biplot, fig.cap="Scatter plot of the first two principal components with observations numerically labelled. The loadings are down by red arrows.", fig.alt="Scatter plot of the second principal component versus the first principal components. Observations as points, numerically labelled. The loadings are shown by red arrows, each labelled by their associated variable."}

Reviewer: ```suggestion
```{r stats-biplot, fig.cap="Scatter plot of the first two principal components with observations numerically labelled. The loadings are shown by red arrows.", fig.alt="Scatter plot of the second principal component versus the first principal component. Observations as points, numerically labelled. The loadings are shown by red arrows, each labelled by their associated variable."}
```

> 
> Let us assume we only care about the principal components accounting for the top
> 80% of the variance in the dataset. Use the `removeVar` argument in `pca()` to remove
> the principal components accounting for the bottom 20%.
> 
> As in the example using prostate data above, examine the first 5 rows and
> columns of rotated data and loadings from your PCA.
> 
> > ## Solution
> > 
> > ```{r pca-ex}

> > > pc <- pca(mat, metadata = metadata) #implement a PCA

Reviewer: ```suggestion
> > pc <- pca(mat, metadata = metadata) # implement a PCA
```

> > #We can check the scree plot to see that many principal components explain a very small amount of
> > the total variance in the data
> > 
> > #Let's remove the principal components with lowest 20% of the variance
> > pc <- pca(mat, metadata = metadata, removeVar = 0.2)
> > #Explore other arguments provided in pca

> > > pc$rotated[1:5, 1:5] #obtain the first 5 principal component scores for the first 5 observations

Reviewer: ```suggestion
> > pc$rotated[1:5, 1:5] # obtain the first 5 principal component scores for the first 5 observations
```

> > pc$loadings[1:5, 1:5] #obtain the first 5 principal component loadings for the first 5 features
> > 

> > > which.max(pc$loadings[, 1]) #index of the maximal loading for the first principal component

Reviewer: ```suggestion
> > which.max(pc$loadings[, 1]) # index of the maximal loading for the first principal component
```

> > > pc$loadings[which.max(pc$loadings[, 1]), ] #principal component loadings for the feature

Reviewer: ```suggestion
> > pc$loadings[which.max(pc$loadings[, 1]), ] # principal component loadings for the feature
```

> > with the maximal loading on the first principal component.
> > 

> > > which.max(pc$loadings[, 2]) #index of the maximal loading for the second principal component

Reviewer: ```suggestion
> > which.max(pc$loadings[, 2]) # index of the maximal loading for the second principal component
```

> > > pc$loadings[which.max(pc$loadings[, 2]), ] #principal component loadings for the feature

Reviewer: ```suggestion
> > pc$loadings[which.max(pc$loadings[, 2]), ] # principal component loadings for the feature
```

> > with the maximal loading on the second principal component.
> > ```

> > > The function `pca()` is used to perform PCA, and uses as inputs a matrix

Reviewer: ```suggestion
> > The function `pca()` is used to perform PCA, and uses as input a matrix
```

> Note that this is different from normalising gene expression data. Gene expression
> data have to be normalised before donwstream analyses can be

> > carried out. This is to reduce to effect technical and other potentially confounding

Reviewer: ```suggestion
> carried out. This is to reduce technical and other potentially confounding
```

> factors. We assume that the expression data we use had been normalised previously.
{: .callout}

## Choosing how many components are important to explain the variance in the data

As in the example using the `prostate` dataset, we can use a scree plot to
compare the proportion of variance in the data explained by each principal
component. This allows us to understand how much information in the microarray
dataset is lost by projecting the observations onto the first few principal
components and whether these principal components represent a reasonable
amount of the variation. The proportion of variance explained should sum to one.

> ## Challenge 4
> 
> Using the `screeplot()` function in **`PCAtools`**, create a scree plot to show 
> proportion of variance explained by each principal component. Explain the

> > output of the scree plot in terms of proportion of variance in data explained

Reviewer: ```suggestion
> output of the scree plot in terms of the proportion of variance in the data explained
```

Finally, it can sometimes be of interest to compare how certain variables contribute
to different principal components. This can be visualised with `plotloadings()` from
the **`PCAtools`** package. The function checks the range of loadings for each
principal component specified (default: first five principal components). It then selects the features
in the top and bottom 5% of these ranges and displays their loadings. This behaviour
can be adjusted with the `rangeRetain` argument, which has 0.1 as the default value (i.e.
5% on each end of the range). NB, if there are too many labels to be plotted, you will see
a warning. This is not a serious problem.

> ```{r loadingsplots, fig.cap=c("Plot of the features in the top and bottom 5 % of the loadings range for the first principal component.", "Plot of the features in the top and bottom 5 % of the loadings range for the second principal component.", fig.cap="Plot of the features in the top and bottom 5 % of the loadings range for the either the first or second principal components."), fig.alt=c("A plot with axes component loading versus principal component index. The plot shows the loadings of features at the top and bottom 5 % of the loadings range and only for the first principal component. The component loading values for four features are shown. The features with the highest loadings are shown at the top of the plot and consist of three features, each with a blue dot and feature label. The feature with the lowest loading is shown at the bottom of the plot and delineated by a yellow dot with a feature label.", "A plot with axes component loading versus principal component index. The plot shows the loadings of features at the top and bottom 5 % of the loadings range and only for the second principal component. The component loading values for 9 features are shown. The features with the highest loadings are shown at the top of the plot and consist of 6 features, each with a blue dot and feature label. The 3 features at the lower range are at the bottom of the plot and delineated by a yellow dot with a feature label.", fig.alt="A plot with axes component loading versus principal component index. The plot shows the loadings of features at the top and bottom 5 % of the loadings range for either principal component. The component loading values for several features are shown with loading score-delineated colours ranging from dark blue for the highest loadings and dark yellow for the lowest loadings. Each point has a feature label. The loadings of features points for the first principal component are concentrated towards the top, while the loadings of features points for the second principal components are concentrated at the top, middle and bottom of the range.")}

Reviewer: ```suggestion
```{r loadingsplots, fig.cap=c("Plot of the features in the top and bottom 5 % of the loadings range for the first principal component.", "Plot of the features in the top and bottom 5 % of the loadings range for the second principal component.", fig.cap="Plot of the features in the top and bottom 5 % of the loadings range for the either the first or second principal components."), fig.alt=c("A plot with component loading versus principal component index. The plot shows the loadings of features at the top and bottom 5 % of the loadings range and only for the first principal component. The component loading values for four features are shown. The features with the highest loadings are shown at the top of the plot and consist of three features, each with a blue dot and feature label. The feature with the lowest loading is shown at the bottom of the plot and delineated by a yellow dot with a feature label.", "A plot with component loadings versus principal component index. The plot shows the loadings of features at the top and bottom 5 % of the loadings range and only for the second principal component. The component loading values for 9 features are shown. The features with the highest loadings are shown at the top of the plot and consist of 6 features, each with a blue dot and feature label. The 3 features at the lower range are at the bottom of the plot and delineated by a yellow dot with a feature label.", fig.alt="A plot with component loading versus principal component index. The plot shows the loadings of features at the top and bottom 5 % of the loadings range for either principal component. The component loading values for several features are shown with loading score-delineated colours ranging from dark blue for the highest loadings and dark yellow for the lowest loadings. Each point has a feature label. The loadings of features points for the first principal component are concentrated towards the top, while the loadings of features points for the second principal components are concentrated at the top, middle and bottom of the range.")}
```
